implement memory utilization fixture

lipxu opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)
27216861

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405

### Approach
#### What is the motivation for this PR?
Test gap for memory leak

#### How did you do it?
Introduced a new fixture to collect the memory information before and after the test case.
Then compare the memory information to confirm that it has not exceeded the high memory usage threshold and that no memory leaks have occurred.

#### How did you verify/test it?
Run case locally.
Set a fake threshold, check the result.
Run nightly pipeline.

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

wsycqyz commented on tests/common/plugins/memory_utilization/memory_utilization_common.json:
```diff
@@ -0,0 +1,15 @@
+      "memory_params": {
+        "memory_usage": {
+          "memory_increase_threshold": 5,
+          "memory_high_threshold": 70
```
Just question. Any reason why it is 70 not 80?

lipxu replied:
Thanks for asking, Currently, the SONiC Memory Utilization Threshold is 70% in ICM. So we set 70 here in public branch.

wsycqyz commented:
/azp run

lipxu commented on tests/common/plugins/memory_utilization/README.md:
```diff
@@ -0,0 +1,88 @@
+          "memory_high_threshold": 70
+        }
+      },
+      "memory_check": "parse_monit_status_output"
```
Thanks for the reminder, updated the README adding more descriptions, thanks

StormLiangMS commented on tests/common/plugins/memory_utilization/README.md:
````diff
@@ -0,0 +1,88 @@
+    ]
+    ```
+
+#### Example of memory items configuration in json file
````
hi @lipxu  could we list the command we recommend for the memory usage / leak check? And how they compare when it is exceeded the threshold with examples?

lipxu replied:
Sure, updated the README with additional examples for current support memory items. thanks

StormLiangMS commented on tests/common/plugins/memory_utilization/README.md:
```diff
@@ -0,0 +1,88 @@
+#### Config files
+Config files, including common file and platform dependence file.
+- **memory_utilization_common.json**: Common configurations for memory utilization.
+- **memory_utilization_dependence.json**: Dependency configruations for memory utilization, currently supporting special configurations for specific HwSku devices. This configuration file should be left empty in the public branch, with special configurations added in internal branch.
```
@lipxu  could we give detailed example about the hwsku configuration?
could we have customized check per testcase?

lipxu replied:
Sure, had updated the README, and add the example for hwsku configuration

StormLiangMS commented:
@lipxu could we have an example to give when this check passed / failed?

lipxu commented:
> @lipxu could we have an example to give when this check passed / failed?

Sure, added the examples for supported memory items.

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Failed<br/>- hook id: trailing-whitespace<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/common/plugins/memory_utilization/README.md<br/><br/>fix end of files.........................................................Failed<br/>- hook id: end-of-file-fixer<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/common/plugins/memory_utilization/README.md<br/><br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Passed<br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

ZhaohuiS commented on tests/common/plugins/memory_utilization/README.md:
```diff
@@ -0,0 +1,573 @@
+  - It includes "COMMON" key which holds all publicly defined common memory items.
+- **memory_utilization_dependence.json**: Dependency configruations for memory utilization, currently supporting special configurations for specific HwSku devices.
+  - The dependency file is also in JSON format and is used in internal branch.
+  - It alos includes a 'COMMON' key which holds all internal common memory items. If an internal memory item has the same name as an item in the public common configuration file, it will overwrite the public common item.
```
@lipxu typo, also

lipxu replied:
thanks for reminder, fixed.

ZhaohuiS commented on tests/common/plugins/memory_utilization/README.md:
````diff
@@ -0,0 +1,573 @@
+
+#### To skip memory_utilization for:
+- all test cases - use pytest command line option ```--memory_utilization```
+- specific test case: mark test case with ```@pytest.mark.memory_utilization``` decorator. Example is shown below.
````
@lipxu should be disable_memory_utilization decorator?

lipxu replied:
thanks for reminder, fixed.

ZhaohuiS commented on tests/common/plugins/memory_utilization/README.md:
````diff
@@ -0,0 +1,573 @@
+- if memory_utilization finds any exceeded thresholds for high memory usage or memory increase, it will display the result and pytest will generate an 'error'.
+
+#### To skip memory_utilization for:
+- all test cases - use pytest command line option ```--memory_utilization```
````
@lipxu a little confusion here, is --memory_utilization  disable or enable the memory check?

lipxu replied:
Thanks for the reminder, fixed. 
memory_utilization , Default is enabled, if we want to disable the check, we can use disable_memory_utilization to skip the memory check.

StormLiangMS commented:
/azp run Azure.sonic-mgmt

StormLiangMS commented on tests/common/plugins/memory_utilization/README.md:
```diff
@@ -0,0 +1,573 @@
+### Scope
```
@lipxu  could we have a section as use case, and give introduction how to use this fixture?

1. To define a memory check globally. 
2. To define a memory check per HWSKU
3. To define a memory check per test module/test case
4. When different memory check defined for global, HWSKU, test cases, how them interact?

lipxu replied:
Sure, updated the README , add "Example Use Cases for memory items" section, thanks for reminder, please review again.

lipxu commented:
/azp run Azure.sonic-mgmt

StormLiangMS commented:
hi @lipxu should we consider when this fixture can kick in? if it is after teardown, it may not work as expected since all process could be already reloaded without memory leak evidence.

lipxu commented:
> hi @lipxu should we consider when this fixture can kick in? if it is after teardown, it may not work as expected since all process could be already reloaded without memory leak evidence.

I've updated the branch by adding the following two functions and adjusted the setup to use "trylast" and teardown to use "tryfirst". This will allow memory utilization to collect the memory information before and after the test case.  ensuring accurate capture memory usage of the test case.

@pytest.hookimpl(trylast=True)
def pytest_runtest_setup(item):

@pytest.hookimpl(tryfirst=True)
def pytest_runtest_teardown(item, nextitem):

StormLiangMS reviewed (approved):
LGTM

StormLiangMS commented:
hi @lipxu LGTM, could you work with other reviewers to close the comment before we get this merged?

dhanasekar-arista commented:
> capture memory usage of the test case

@StormLiangMS  @lipxu  I have a concern on this block 
```
@pytest.hookimpl(trylast=True)
def pytest_runtest_setup(item):

@pytest.hookimpl(tryfirst=True)
def pytest_runtest_teardown(item, nextitem):
```

There may be fixtures managing configuration setup and cleanup around the test. In such cases, the memory checker should run only after all post-test cleanup has been completed. Else memory checker wouldn't give us expected results.
